### ISSUE TITLE
EPUB conversion still not quite to standard for iBooks Proofs

### ISSUE THREAD

**Issue by sbaker7** [jgm/pandoc#2456]
Hi all,

I use pandoc to convert to epub from latex quite regularly and came across an issue where iBooks will not accept the epubs as proof documents (which allows me to sync them with my iPhone/iPad for when I proof read them and make comments etc).

The issue (from what I can tell) seems to be with the compression. By using an applescript tool I found online, I have a feeling the reason why is because of the amount of compression on the mimetype and DS_STORE files. Mimetype files are not allowed to be compressed at all under EPUB 3.0 standards.

The fix, at least on OSX, is to run these commands in terminal:

`zip -X0 "FilePath" mimetype`
`zip -rDX9 "FilePath" * -x "*.DS_Store" -x mimetype`

Basically, don't apply any compression to the mimetype or DS_STORE files. Although this fix works (most of the time), it's annoying to run after every compilation on every epub each time I compile. Hopefully the stuff I've provided helps and it's a minor fix; I'd try and fix it myself but I haven't got time to sit down and learn Haskell at the moment!

Is there any chance this will be fixed any time soon (if at all), particularly as the website does advertise pandoc supports EPUB 3 and (technically) this means it actually doesn't?

**Comment by jgm** [comment]
The algorithm pandoc uses to create the epub (zip) containers only compresses files when there's an actual space savings from doing so.  With the 20 byte `mimetype` file there isn't, so it doesn't get compressed.  (I just verified that by using `unzip -v` on a pandoc-created EPUB; it shows that the content was stored and did not need to be deflated.)  So you should not be seeing a compressed `mimetype` file.  You also shouldn't be seeing a `.DS_Store` at all in your zip container.  Pandoc never inserts one.

I am therefore skeptical that the file you're dealing with was actually created by pandoc.

Could you use pandoc to create a simple EPUB, and send or link to the resulting file plus the exact input source and pandoc command you used to create it -- and also the pandoc version you used?

**Comment by jgm** [comment]
I'd love to resolve this issue one way or another.  Since I've never seen a pandoc-produced epub where the mimetype file is compressed, please give me a way to reproduce this problem locally.

**Comment by sbaker7** [comment]
Sorry for the delay, I meant to reply to this earlier. Here's a [link](https://www.dropbox.com/sh/x1mczokbyekmnop/AABCpwk0k8kofhuHCXNJNqtua?dl=0&s=so) to a folder on dropbox, which has a basic latex document and ePub that I generated with the following command:

`pandoc FamilyPleas.tex -t EPUB3 -o FamilyPleas.epub`

By using the fix above (unzipping, rezipping with no compression), it suddenly becomes valid for use with iBooks as a proof document. I may have missed something else, but that seemed to be the solution with the limited understanding of the inner workings of pandoc I have.

**Comment by jgm** [comment]
Using `unzip -v` to view the epub archive, we can see that (i) pandoc does not create a .DS_Store file in the archive, and (ii) the mimetype file is not compressed in the archive.  (Method is "stored.")
```
jgm@MacBook-Air:~/Dropbox/EPUBTest/Family Pleas % pandoc FamilyPleas.tex -t EPUB3 -o F.epub
jgm@MacBook-Air:~/Dropbox/EPUBTest/Family Pleas % unzip -v F.epub
Archive:  F.epub
 Length   Method    Size  Ratio   Date   Time   CRC-32    Name
--------  ------  ------- -----   ----   ----   ------    ----
      20  Stored       20   0%  10-21-15 05:04  2cab616f  mimetype
     246  Defl:X      166  33%  10-21-15 05:04  e8e295ae  META-INF/container.xml
     160  Defl:X      119  26%  10-21-15 05:04  d2f87677  META-INF/com.apple.ibooks.display-options.xml
     576  Defl:X      276  52%  10-21-15 05:04  afd1e7c1  stylesheet.css
     465  Defl:X      286  39%  10-21-15 05:04  c7fd9da0  title_page.xhtml
    2748  Defl:X      685  75%  10-21-15 05:04  4e704791  content.opf
    7213  Defl:X     1427  80%  10-21-15 05:04  ed2a0356  toc.ncx
    4808  Defl:X     1375  71%  10-21-15 05:04  1f11a91d  nav.xhtml
   27567  Defl:X    26811   3%  10-21-15 05:04  d88f2408  media/file2.png
   18009  Defl:X    17316   4%  10-21-15 05:04  bae7da9d  media/file1.png
    7978  Defl:X     7004  12%  10-21-15 05:04  99c3be4d  media/file0.png
    2006  Defl:X     1088  46%  10-21-15 05:04  64c85166  ch001.xhtml
    2086  Defl:X     1126  46%  10-21-15 05:04  c1e6e43e  ch002.xhtml
    1291  Defl:X      721  44%  10-21-15 05:04  9e4d6608  ch003.xhtml
    1272  Defl:X      723  43%  10-21-15 05:04  bb54f9be  ch004.xhtml
    2550  Defl:X     1280  50%  10-21-15 05:04  dc3a2ac3  ch005.xhtml
    1677  Defl:X      934  44%  10-21-15 05:04  5b7d4920  ch006.xhtml
    2823  Defl:X     1361  52%  10-21-15 05:04  f71648c0  ch007.xhtml
    2108  Defl:X     1105  48%  10-21-15 05:04  bb597387  ch008.xhtml
    1546  Defl:X      721  53%  10-21-15 05:04  3a294951  ch009.xhtml
    1789  Defl:X      732  59%  10-21-15 05:04  9246617f  ch010.xhtml
    3649  Defl:X     1653  55%  10-21-15 05:04  006804b8  ch011.xhtml
    1494  Defl:X      842  44%  10-21-15 05:04  3ffd7c22  ch012.xhtml
--------          -------  ---                            -------
   94081            67771  28%                            23 files
```
So I don't think your diagnosis is correct.  But maybe something else is wrong about these epubs?  Let's try epubcheck:
```
% epubcheck F.epub 
Epubcheck Version 3.0.1

Validating against EPUB version 3.0
ERROR: F.epub/content.opf(6,39): character content of element "dc:date" invalid; must be a string with length at least 1 (actual length was 0)
WARNING: F.epub/content.opf(6,39): date value '' does not follow recommended syntax as per http://www.w3.org/TR/NOTE-datetime: zero-length string
ERROR: F.epub/title_page.xhtml(12,24): element "p" not allowed here; expected the element end-tag, text or element "a", "abbr", "area", "audio", "b", "bdi", "bdo", "br", "button", "canvas", "cite", "code", "command", "datalist", "del", "dfn", "em", "embed", "epub:switch", "i", "iframe", "img", "input", "ins", "kbd", "keygen", "label", "map", "mark", "meter", "ns1:math", "ns2:svg", "object", "output", "progress", "q", "ruby", "s", "samp", "script", "select", "small", "span", "strong", "sub", "sup", "textarea", "time", "u", "var", "video" or "wbr" (with xmlns:ns1="http://www.w3.org/1998/Math/MathML" xmlns:ns2="http://www.w3.org/2000/svg")

Check finished with warnings or errors
```
So, several problems here.  You need a valid date (and it has to be in a particular format).

There's also an issue with the title.  Pandoc is parsing the `title` metadata as a Para (paragraph), rather than a Plain.  That's probably because of the `\vspace` commands, which make this a sequence of block elements rather than inlines.  And HTML doesn't like the p tag inside an h1.  With this issue, we might want to modify the HTML writer to convert the title to a sequence of Inlines, no matter what its structure.

**Comment by sbaker7** [comment]
Okay, I have come across those before. This is just one of many files I make, I had another I had fixed all those issues in but it's huge, so I didn't want to share it. The issue still remained the same.

**Comment by jgm** [comment]
When I run
```
% pandoc -t epub3 George.tex -o G.epub --number-sections
% epubcheck G.epub
```
I get
```
Validating against EPUB version 3.0
No errors or warnings detected.
```
So this seems to be a perfectly conforming EPUB3, unless there's a bug in epubcheck 3.0.1.

**Comment by sbaker7** [comment]
Yeah, so I'm getting that, but it's still not actually conforming correctly (which may be an epubcheck issue) because it won't load into iBooks as a proof file (which the requirements for is purely that it must be epub 3.0). The only way I can make that work is to run it through a script with the commands I specified earlier, so something has to be not quite right (or iBooks secretly has some other requirement that I haven't been able to identify)

**Comment by jgm** [comment]
So I can try this: What do I do to try to load an epub into iBooks as a "proof file"?

**Comment by sbaker7** [comment]
Directions can be found on apple’s website here: [link](https://support.apple.com/en-au/HT204531)

From what I can tell with their directions, they're basically saying is 'make it a folder, to then make it an ePub again'. Again, I may have missed a step in something that's going on in the background but something doesn't quite make sense to me

**Comment by jgm** [comment]
I followed the instructions there, and it worked perfectly.
Here's how to do it on the command line:

```
pandoc George.tex -o George.epub --number-sections
mkdir GeorgeProof.epub
cd GeorgeProof.epub
unzip ../GeorgeProof.epub
```
Now go to the Advanced Menu in iBooks, open as proof, and select the folder GeorgeProof.epub.
It opens with no errors or warnings.

Note:  a proof *is* a folder, not a zip file!

Upshot: no problem here with pandoc.
